Inheritance

GintaPerep opened the pull request.

GintaPerep committed:
feat: created ES6 class IntBuilder

GintaPerep committed:
fix: clean code and structur

GintaPerep committed:
feat: first functions in Builder_ES5

GintaPerep committed:
fix: changed structure of code

GintaPerep committed:
feat: completed code for builder_ES5

matveyprupas commented:
Hi Ginta, 

I found a few necessary improvements that you can make before the deadline :)

1. Both ES5 and ES6 methods are not chainable. For example, I can't start the next code: `intBuilder.plus(3).get();`.  
2. ES5 methods you implemented your own methods with prohibited methods (String.prototype.repeat() and String.prototype.replace()). 

Change your implementation according to these points above, please.

matveyprupas commented on inheritance/builder_ES6.js:
```diff
@@ -0,0 +1,66 @@
+class Builder {
+    constructor(_initNum = 0) {
+       this.initNum = _initNum;
+       this.sum = 0;
```
Why do we need this additional variable `sum`? 
Maybe you can implement only with `initNum`?

matveyprupas commented on inheritance/builder_ES6.js:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    minus(...n) {
+        for (let arg of n) this.sum -= arg;
```
What will you see in a console as a result of the next code? 

```
const intBuilder = new IntBuilder(10);
intBuilder.minus(1,2)
intBuilder.plus(2,3,2)
```

Believe that we need to find a solution without the variable `this.sum`.

matveyprupas commented on inheritance/builder_ES5.js:
```diff
@@ -0,0 +1,67 @@
+  
+  function BuilderES5 (greeting) {
+    this.greeting = greeting
+    this.text = ''
```
The same comment

Why do we need this additional variable `text`?
Maybe you can implement only with `greeting`?

matveyprupas commented on inheritance/builder_ES5.js:
```diff
@@ -0,0 +1,67 @@
+  }
+
+  BuilderES5.prototype.minus = function (n) {
+    this.text = this.text.slice(0, - n)
```
The same comment.

What will you see in a console as a result of the next code?

```
let strBuilder = new StringBuilder('Hello');
strBuilder.minus(4)
strBuilder.plus(' all','!')
```

Believe that we need to find a solution without the variable `this.text`.

matveyprupas commented on inheritance/builder_ES5.js:
```diff
@@ -0,0 +1,67 @@
+  }
+
+  BuilderES5.prototype.remove = function (str) {
+    this.text = this.text.replaceAll(str, '')
```
Resolve it without prohibited method `replaceAll`

matveyprupas commented on inheritance/builder_ES5.js:
```diff
@@ -0,0 +1,67 @@
+  }
+
+  BuilderES5.prototype.multiply = function (int) {
+   this.text = this.text.repeat(int)
```
Resolve it without prohibited method `repeat`

matveyprupas commented:
And an additional comment regarding clean code. 

You've added in every method `console.log`. 
`console.log` could help us during creating code, but in production, it will be annoying to see a lot of information in the console. Better to use `console.log` in one place at the end of your code. It's easier to remove it before production.

For example, like this:

```
let strBuilder = new StringBuilder('Hello');

console.log(
    strBuilder
        .plus(' all','!')
        .minus(4)
        .get()
);
```

For our task, it is not necessary to remove it, but in future work, this approach will be more helpful.

GintaPerep committed:
fix: relace() method in ES5

GintaPerep committed:
fix: code refactoring

GintaPerep commented:
> And an additional comment regarding clean code.
> 
> You've added in every method `console.log`. `console.log` could help us during creating code, but in production, it will be annoying to see a lot of information in the console. Better to use `console.log` in one place at the end of your code. It's easier to remove it before production.
> 
> For example, like this:
> 
> ```
> let strBuilder = new StringBuilder('Hello');
> 
> console.log(
>     strBuilder
>         .plus(' all','!')
>         .minus(4)
>         .get()
> );
> ```
> 
> For our task, it is not necessary to remove it, but in future work, this approach will be more helpful.

Thank you for your comment and advise.

GintaPerep commented:
Hi,
1. repaired,
2. repaired.
Can you please check?

Thank you!

matveyprupas commented on inheritance/builder_ES5.js:
```diff
@@ -0,0 +1,77 @@
+  }
+
+  BuilderES5.prototype.multiply = function (int) {
+   this.text = Array(int + 1).join(this.text);
```
Cool! Highly original solution!

matveyprupas reviewed (changes requested):
## ES5: 

1. remove method doesn't work correctly. It removes all symbols from first "l". 
The expected result is: "Heo a!", but actually result: "He" 
It needs improvement.  **Changes are required.**

<img width="509" alt="Screenshot 2023-01-02 at 12 59 18" src="https://user-images.githubusercontent.com/60707739/210222605-fb13a2a1-1321-4114-a418-303736872f28.png">

2. In case I remove method plus all next methods will return ''. 
It is not correct. You get this error since you use accumulator `this.text`. Actually, you don't need it. Try to resolve it without this additional variable. **Changes are required.**

<img width="527" alt="Screenshot 2023-01-02 at 13 04 28" src="https://user-images.githubusercontent.com/60707739/210223061-b769c6d6-49b1-41ab-bc78-21b88a9775ad.png">

3. Btw, I found one more mistake in method `divide`. It should contain variable `k = Math.floor(str.length / n);` But you use it without this variable. **Changes are required.**


## ES6

1. The same situation with `this.sum`. Find the solution without it. In the case without method `plus` I see in the console next situation: `10-3 = -3` See screenshot below. 
**Changes are required.**

<img width="511" alt="Screenshot 2023-01-02 at 13 20 19" src="https://user-images.githubusercontent.com/60707739/210224516-da4f3760-7b20-4dcd-ac25-715be3d731ec.png">

GintaPerep committed:
fix: I fixed the code structure (removed t

GintaPerep commented:
Hi Matvey,
1. I fixed the code structure, removed this.text and this.sum;
2. Fixed  variable k = Math.floor(str.length / n);
3. Unfortunately, I did see that chainable functions calls can be any order in task descriptions. If this point is mandatory for the task, I would like to have more time for improving my code.

Thanks a lot!

GintaPerep committed:
fix: fixed remove function

GintaPerep commented:
Hi,
Looks like I fixed remove function. :)
